### PR TITLE
implementing two-phase grain growth model

### LINKED ISSUE #8837: generalizing the single-phase grain growth model to two-phase systems
### Description of the enhancement or error report

Extending the single-phase grain growth models to two-phase systems, where each phase can have multiple orientations based on Moelans model (Moelans Acta Mat 2011). 

### Rationale for the enhancement or information for reproducing the error
It can be used for investigating grain growth in two-phase systems.  This is also an important step for multi-phase phase-field models, where multiple species concentrations can also be coupled to that model.  

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by karim-E-ahmed**
closes #8837

### CODE REVIEW

**Comment by karim86** [comment]
@dschwen I had to close the old PR and open a new one (basically I created a new branch) for some strange behavior with git push

File: modules/phase_field/include/action/PolycrystalKernelAction.h [hunk 1]
```diff
@@ -21,6 +21,9 @@ class PolycrystalKernelAction : public Action
   virtual void act();
 
 protected:
+  /// number of phases
+  unsigned int _phase_num;
```
Comment (dschwen) [inline modules/phase_field/include/action/PolycrystalKernelAction.h]: No, this only belongs in the derived class!

File: modules/phase_field/src/action/PolycrystalKernelAction.C [hunk 2]
```diff
@@ -40,6 +43,13 @@ PolycrystalKernelAction::PolycrystalKernelAction(const InputParameters & params)
 void
 PolycrystalKernelAction::act()
 {
+  std::string kernel_type;
+  if (_phase_num == 1)
+    kernel_type = "ACGrGrPoly";
+  else if (_phase_num == 2)
+    kernel_type = "TwoPhaseACGrGrPoly";
```
Comment (dschwen) [inline modules/phase_field/src/action/PolycrystalKernelAction.C]: No, this logic needs to be implemented using a virtual name function

File: modules/phase_field/include/action/PolycrystalKernelAction.h (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/include/action/PolycrystalKernelAction.h]: If this is done right this member is not needed at all

File: modules/phase_field/src/action/PolycrystalKernelAction.C [hunk 3]
```diff
@@ -60,16 +70,18 @@ PolycrystalKernelAction::act()
     //
 
     {
-      InputParameters params = _factory.getValidParams("ACGrGrPoly");
+      InputParameters params = _factory.getValidParams(kernel_type);
       params.set<NonlinearVariableName>("variable") = var_name;
```
Comment (dschwen) [inline modules/phase_field/src/action/PolycrystalKernelAction.C]: use `applyParameters` to simplify this

**Review by dschwen — CHANGES_REQUESTED** [review]
The design is not right yet

File: modules/phase_field/src/action/PolycrystalTwoPhaseKernelAction.C [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  params.addParam<Real>("mob_ratio", 1.0, "Ratio of surface energy to GB mobility");
+  return params;
+}
+
+PolycrystalTwoPhaseKernelAction::PolycrystalTwoPhaseKernelAction(const InputParameters & params)
+  : PolycrystalKernelAction(params),
+    _second_phase_op_num(getParam<unsigned int>("second_phase_op_num")),
+    _en_ratio(getParam<Real>("en_ratio")),
+    _second_phase_en_ratio(getParam<Real>("second_phase_en_ratio")),
+    _mob_ratio(getParam<Real>("mob_ratio"))
```
Comment (dschwen) [inline modules/phase_field/src/action/PolycrystalTwoPhaseKernelAction.C]: all these members can be removed!

File: modules/phase_field/src/kernels/ACTwoPhaseGrGrPoly.C [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+ACTwoPhaseGrGrPoly::ACTwoPhaseGrGrPoly(const InputParameters & parameters)
+  : ACGrGrBase(parameters),
+    _gamma(getMaterialProperty<Real>("gamma_asymm")),
+    _op_index(getParam<unsigned int>("op_index")),
+    _second_phase_op_num(getParam<unsigned int>("second_phase_op_num")),
+    _en_ratio(getParam<Real>("en_ratio")),
+    _second_phase_en_ratio(getParam<Real>("second_phase_en_ratio")),
+    _mob_ratio(getParam<Real>("mob_ratio"))
+
```
Comment (dschwen) [inline modules/phase_field/src/kernels/ACTwoPhaseGrGrPoly.C]: rm line

**Review by dschwen — CHANGES_REQUESTED** [review]
I cannot merge a new model without a .md file to document it (with equations). Have you used the new documentation system at all? I can show you.

**Comment by karim86** [comment]
I have not used yet, I'll stop by

On Thu, Apr 27, 2017 at 3:52 PM, Daniel Schwen <notifications@github.com>
wrote:

> *@dschwen* requested changes on this pull request.
>
> I cannot merge a new model without a .md file to document it (with
> equations). Have you uses the new documentation system at all? I can show
> you.
> ------------------------------
>
> In modules/phase_field/src/action/PolycrystalTwoPhaseKernelAction.C
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_8989-23discussion-5Fr113811317&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=KOzEFeajphUNNJjdyCR7w39YSjM6k_ww-KKJ-3fNQJg&s=75v7U8xqoegl5xIBRKK5E9odKBdSFOiAkshxaSEjloU&e=>
> :
>
> > +      1,
> +      "specifies the total number of grains/variants of the second phase");
> +  params.addParam<Real>(
> +      "en_ratio", 1.0, "Ratio of surface energy to GB energy, e.g., interphase energy");
> +  params.addParam<Real>(
> +      "second_phase_en_ratio", 1.0, "Ratio of second-phase to parent-phase GB energy");
> +  params.addParam<Real>("mob_ratio", 1.0, "Ratio of surface energy to GB mobility");
> +  return params;
> +}
> +
> +PolycrystalTwoPhaseKernelAction::PolycrystalTwoPhaseKernelAction(const InputParameters & params)
> +  : PolycrystalKernelAction(params),
> +    _second_phase_op_num(getParam<unsigned int>("second_phase_op_num")),
> +    _en_ratio(getParam<Real>("en_ratio")),
> +    _second_phase_en_ratio(getParam<Real>("second_phase_en_ratio")),
> +    _mob_ratio(getParam<Real>("mob_ratio"))
>
> all these members can be removed!
> ------------------------------
>
> In modules/phase_field/src/kernels/ACTwoPhaseGrGrPoly.C
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_8989-23discussion-5Fr113811355&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=KOzEFeajphUNNJjdyCR7w39YSjM6k_ww-KKJ-3fNQJg&s=3PTY9CpJ2XTwR0EnZz7Y1ZsU6YsDb2IUczJJNcA8hxU&e=>
> :
>
> > +      "en_ratio", 1.0, "Ratio of surface energy to GB energy, e.g., interphase energy");
> +  params.addParam<Real>(
> +      "second_phase_en_ratio", 1.0, "Ratio of second-phase to parent-phase GB energy");
> +  params.addParam<Real>("mob_ratio", 1.0, "Ratio of surface energy to GB mobility");
> +  return params;
> +}
> +
> +ACTwoPhaseGrGrPoly::ACTwoPhaseGrGrPoly(const InputParameters & parameters)
> +  : ACGrGrBase(parameters),
> +    _gamma(getMaterialProperty<Real>("gamma_asymm")),
> +    _op_index(getParam<unsigned int>("op_index")),
> +    _second_phase_op_num(getParam<unsigned int>("second_phase_op_num")),
> +    _en_ratio(getParam<Real>("en_ratio")),
> +    _second_phase_en_ratio(getParam<Real>("second_phase_en_ratio")),
> +    _mob_ratio(getParam<Real>("mob_ratio"))
> +
>
> rm line
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_8989-23pullrequestreview-2D35233398&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=KOzEFeajphUNNJjdyCR7w39YSjM6k_ww-KKJ-3fNQJg&s=BaOS6cXfzDhsEp1LaJYbTHC5mbxxfyDcbnGKOi1kwjA&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AOW-5F85CcrHx8amxNbJHFjvjQ-2D6NSlrz8ks5r0Q4mgaJpZM4NEldh&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=KOzEFeajphUNNJjdyCR7w39YSjM6k_ww-KKJ-3fNQJg&s=8jg0pmk9f1zK6s6eOKkLlr9irL7WRmzmntGUyKCSjKE&e=>
> .
>



-- 
Karim El-Sayed Ahmed, PhD
Postdoctoral Research Associate
Fuel Modeling & Simulation Department
Idaho National Laboratory

**Comment by permcody** [comment]
#9145 is incompatible with this PR. Depending on which one goes in, we'll have to update this PR. The new system is already merged though so these new tests can be updated. @karim86 - let me know if need help.

**Comment by karim86** [comment]
Yeah I am going to rebase anyway to add the documentation. I will look at the new system and see how to modify the test. I will let you guys know if I needed help.

**Comment by dschwen** [comment]
@karim86 do you need any help with this?

**Comment by karim86** [comment]
@dschwen thanks Daniel I think I got it I just got busy working on different stuff last couple of weeks. I should finish this soon.

**Comment by dschwen** [comment]
Ping @karim86

**Comment by dschwen** [comment]
@karim86 has not reacted on here for 47 days. Policy is to close after 30days of inactivity. I'd rather see this get finished though...

**Comment by permcody** [comment]
@dschwen - With @karim86 gone this is going to remain open indefinitely unless somebody else is assigned. What would you like to do with this PR?

**Comment by karim86** [comment]
I will update it in the next few days.

**Comment by permcody** [comment]
Closing due to inactivity.

**Pull request closed without merging**